src/emqx_logger_formatter.erl

```diff
@@ -0,0 +1,360 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 2017-2018. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+
+%% This file is copied from lib/kernel/src/logger_formatter.erl, and
+%% modified for a more concise time format other than the default RFC3339.
+
+-module(emqx_logger_formatter).
+
+-export([format/2]).
+-export([check_config/1]).
+
+-define(DEFAULT_FORMAT_TEMPLATE_SINGLE, [time," ",level,": ",msg,"\n"]).
+
+-define(FormatP, "~0tp").
+
+-define(IS_STRING(String),
+        (is_list(String) orelse is_binary(String))).
+
+%%%-----------------------------------------------------------------
+%%% Types
+-type config() :: #{chars_limit     => pos_integer() | unlimited,
+                    depth           => pos_integer() | unlimited,
+                    max_size        => pos_integer() | unlimited,
+                    report_cb       => logger:report_cb(),
+                    quit        => template()}.
+-type template() :: [metakey() | {metakey(),template(),template()} | string()].
+-type metakey() :: atom() | [atom()].
+
+%%%-----------------------------------------------------------------
+%%% API
+-spec format(LogEvent,Config) -> unicode:chardata() when
+      LogEvent :: logger:log_event(),
+      Config :: config().
+format(#{level:=Level,msg:=Msg0,meta:=Meta},Config0)
+  when is_map(Config0) ->
+    Config = add_default_config(Config0),
+    Template = maps:get(template,Config),
+    {BT,AT0} = lists:splitwith(fun(msg) -> false; (_) -> true end, Template),
+    {DoMsg,AT} =
+        case AT0 of
+            [msg|Rest] -> {true,Rest};
+            _ ->{false,AT0}
+        end,
+    B = do_format(Level,Meta,BT,Config),
+    A = do_format(Level,Meta,AT,Config),
+    MsgStr =
+        if DoMsg ->
```

Comment:
this whole block from this line to line 80 worth an abstraction to a help function.

Reply:
Agree. But this whole file is copied from OTP `lib/kernel/src/logger_formatter.erl` maily for new date format, and I'd like to keep most of the file untouched.